### DISCUSSION TITLE
Dev mode not working (for SPA) with `base`/`basename` set

### DISCUSSION THREAD

**Discussion by dimw** [remix-run/remix#9375] [Q&A] [answered]
I'm trying to run an SPA with the base set to `/app/`. Therefore, I'm following the [Remix' SPA Mode doc](https://remix.run/docs/en/main/guides/spa-mode) as well as the [Deploying Remix to GitHub Pages tutorial](https://youtu.be/EO_-KwSslWQ?t=451) from @brookslybrand.

The steps are:
1. Bootstrapping a project from a template:
   ```sh
   npx create-remix@latest --template remix-run/remix/templates/spa spa-test
   ```
2. Changing the `base` and `basename` to `/app/` (to make accessible the SPA from http://example.com/app/):
   ```ts
   // vite.config.ts
   export default defineConfig({
     base: "/app/",   // <-- This line is new
     plugins: [
       remix({
         ssr: false,
         basename: "/app/",  // <-- This line is new
       }),
       tsconfigPaths(),
     ],
   });

When I now build (`npm run build`) and execute the preview (`npm run preview`), the page is working as expected under http://localhost:4173/app/.

### Issue
However, when running in the dev mode (`npm run dev`) and opening the provided URL (http://localhost:5173/app/), the dev server retrieves an internal server error while I'm seeing the following error messages on the console.
```shell
[vite] Pre-transform error: Failed to load url /root.tsx (resolved id: /root.tsx) in virtual:remix/server-build. Does the file exist?
[vite] Pre-transform error: Failed to load url /routes/foo.tsx (resolved id: /routes/foo.tsx) in virtual:remix/server-build. Does the file exist?
[vite] Pre-transform error: Failed to load url /routes/_index.tsx (resolved id: /routes/_index.tsx) in virtual:remix/server-build. Does the file exist?
[vite] Error when evaluating SSR module virtual:remix/server-build: failed to import "/root.tsx"
|- Error: Failed to load url /root.tsx (resolved id: /root.tsx) in virtual:remix/server-build. Does the file exist?
    at loadAndTransform (file:///<redacted>/spa-test/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:53884:21)
    at async instantiateModule (file:///<redacted>/spa-test/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:54954:10)

[vite] Internal server error: Failed to load url /root.tsx (resolved id: /root.tsx) in virtual:remix/server-build. Does the file exist?
```

The same error is happening with _dev_ when a regular Remix app (with SSR enabled) is used.

I tried to remove the Vite [`base` option](https://vitejs.dev/config/shared-options.html#base) but this leads to assets being loaded from the root instead of `/app/`. This would require more effort in server config when bringing the app to a server and I would like to avoid that.

<img width="857" alt="image" src="https://github.com/remix-run/remix/assets/2025688/d78fc3b0-33cd-4236-9c31-a1616dde1dce">

### Workaround
For now, I ended up having two Vite configurations:
- Default (dev) config (_vite.config.ts_) with no `base` set
- Build config (_vite.config.build.ts) which is extending the default config and setting the `base` on top:
   ```ts
   // vite.config.build.ts
   export default mergeConfig(defaultConfig, {
     base: "/app/",
   });
   ```
   Which is then executed from _package.json_ as following:
   ```js
   {
     "scripts": {
       "build": "remix vite:build --config ./vite.config.build.ts",
       // ...
     }
   }
   ```

Does anyone have a hint what I'm missing here in the Vite config or is that a bug?

**Versions**
- @remix-run/dev: 2.9.1
- vite: 5.1.0

**Related to**
- https://github.com/remix-run/remix/pull/8145

**Comment by dimw** [comment, marked as answer]
Found it. It looks like the issue appears only for the `base`/`basename` with the value `/app/`. 🤦 For every other values it is working fine. I assume that Vite/Remix plugin interferes with the default Remix source folder name. Will raise an issue.

**Reply by brophdawg11** [reply]
Ah thanks for narrowing this down - that does look like an issue in our vite plugin.  I converted this into an issue in https://github.com/remix-run/remix/issues/9382 for us to look into.

**Reply by lwhiteley** [reply]
Also ran into this issue for ssr and had to update my config to change the appDirectory

```
export default defineConfig({
  base: '/app/',
  plugins: [
    remix({
      appDirectory: 'src',
      basename: '/app/',
    }),
  ],
});

```

also had to update the tconfig references to point to the new location of the `src` files instead of `app/**`
